diff --git a/tute08/src/fs/model/Directory.java b/tute08/src/fs/model/Directory.java
index 1e286df..39a2386 100644
--- a/tute08/src/fs/model/Directory.java
+++ b/tute08/src/fs/model/Directory.java
@@ -1,5 +1,67 @@
- public class Directory
+ public class Directory extends Inode {
# now we know this type
-    private List<???> innerInodes = new ArrayList<>();
+    private List<Inode> innerInodes = new ArrayList<>();

     public Directory(Path path) {
+        super(path);
         this.innerInodes = new ArrayList<>();
     }
 
+    @Override
     public long getFileSize() {
         return super.getFileSize() + innerInodes.stream().mapToLong(i -> i.getFileSize()).sum();
     }

+    @Override
     public String getIcon(boolean isExpanded) {
         return isExpanded ? "folder-open.png" : "folder.png";
     }
 
+    @Override
     public String getName() {
         return Paths.get(getPath()).toString();
     }
 }
